### PR TITLE
fix(custom-transform): allow to assert empty program for rsc

### PR DESCRIPTION
**Pull request by kwonoj**
### What?
Turbopack uses `parse_as_prorgram`, and in case of empty file it returns script instead of module so none of visitor runs for the rsc module.

This PR attempts to workaround, however need double clarification if upstream swc behavior is desired.

Closes PACK-2460

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix-empty-program-rsc-validation | Change |
| - | - | - | - |
| buildDuration | 11.8s | 11.8s | N/A |
| buildDurationCached | 6.2s | 5.4s | N/A |
| nodeModulesSize | 196 MB | 196 MB | ✓ |
| nextStartRea..uration (ms) | 435ms | 426ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix-empty-program-rsc-validation | Change |
| - | - | - | - |
| 3f784ff6-HASH.js gzip | 53.5 kB | 53.5 kB | N/A |
| 423.HASH.js gzip | 185 B | 181 B | N/A |
| 68-HASH.js gzip | 29.6 kB | 29.6 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 238 B | 239 B | N/A |
| main-HASH.js gzip | 31.9 kB | 31.9 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | ✓ |
| Overall change | 46.9 kB | 46.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix-empty-program-rsc-validation | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix-empty-program-rsc-validation | Change |
| - | - | - | - |
| _app-HASH.js gzip | 194 B | 195 B | N/A |
| _error-HASH.js gzip | 182 B | 181 B | N/A |
| amp-HASH.js gzip | 502 B | 501 B | N/A |
| css-HASH.js gzip | 320 B | 322 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 255 B | 256 B | N/A |
| head-HASH.js gzip | 350 B | 349 B | N/A |
| hooks-HASH.js gzip | 368 B | 369 B | N/A |
| image-HASH.js gzip | 4.19 kB | 4.18 kB | N/A |
| index-HASH.js gzip | 257 B | 256 B | N/A |
| link-HASH.js gzip | 2.67 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 310 B | 311 B | N/A |
| script-HASH.js gzip | 384 B | 383 B | N/A |
| withRouter-HASH.js gzip | 306 B | 308 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 106 B | 106 B | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix-empty-program-rsc-validation | Change |
| - | - | - | - |
| _buildManifest.js gzip | 483 B | 485 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix-empty-program-rsc-validation | Change |
| - | - | - | - |
| index.html gzip | 528 B | 527 B | N/A |
| link.html gzip | 541 B | 539 B | N/A |
| withRouter.html gzip | 523 B | 522 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix-empty-program-rsc-validation | Change |
| - | - | - | - |
| edge-ssr.js gzip | 94.4 kB | 94.4 kB | N/A |
| page.js gzip | 150 kB | 150 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix-empty-program-rsc-validation | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 622 B | 626 B | N/A |
| middleware-r..fest.js gzip | 151 B | 149 B | N/A |
| middleware.js gzip | 47.4 kB | 47.4 kB | N/A |
| edge-runtime..pack.js gzip | 1.94 kB | 1.94 kB | ✓ |
| Overall change | 1.94 kB | 1.94 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix-empty-program-rsc-validation | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 166 kB | 166 kB | ✓ |
| app-page-exp..prod.js gzip | 95.4 kB | 95.4 kB | ✓ |
| app-page-tur..prod.js gzip | 97.2 kB | 97.2 kB | ✓ |
| app-page-tur..prod.js gzip | 91.6 kB | 91.6 kB | ✓ |
| app-page.run...dev.js gzip | 136 kB | 136 kB | ✓ |
| app-page.run..prod.js gzip | 90.2 kB | 90.2 kB | ✓ |
| app-route-ex...dev.js gzip | 22 kB | 22 kB | ✓ |
| app-route-ex..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| app-route-tu..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| app-route-tu..prod.js gzip | 14.7 kB | 14.7 kB | ✓ |
| app-route.ru...dev.js gzip | 21.7 kB | 21.7 kB | ✓ |
| app-route.ru..prod.js gzip | 14.7 kB | 14.7 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.43 kB | 9.43 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.7 kB | 9.7 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.43 kB | 9.43 kB | ✓ |
| pages-turbo...prod.js gzip | 22 kB | 22 kB | ✓ |
| pages.runtim...dev.js gzip | 22.7 kB | 22.7 kB | ✓ |
| pages.runtim..prod.js gzip | 22 kB | 22 kB | ✓ |
| server.runti..prod.js gzip | 49.9 kB | 49.9 kB | ✓ |
| Overall change | 924 kB | 924 kB | ✓ |

</details>

</details>
Commit: c3211975d1dd6cf6df609f04de8656b0a60b9ff3

**Comment by sokra** [comment]
Should we fix that in Turbopack to always use a `Module` and avoid `Script`s?

**Comment by kwonoj** [comment]
> Should we fix that in Turbopack to always use a Module and avoid Scripts?

Yes, about to ask kdy to confirm actually - didn't ping him since it's holiday. PR shoudn't be checked in until then.

File: packages/next-swc/crates/next-custom-transforms/src/transforms/react_server_components.rs [hunk 1]
```diff
@@ -727,6 +728,15 @@ impl ReactServerComponentValidator {
 impl Visit for ReactServerComponentValidator {
     noop_visit_type!();
 
+    // Workaround for swc's parse_program, where it detects empty file as a script.
+    // we may fix upstream to autodetect empty file as a module
```
Comment (kdy1) [inline packages/next-swc/crates/next-custom-transforms/src/transforms/react_server_components.rs:732]: Empty file is a script, and you can try it from vscode by importing an empty file from another file.
vscode will say `foo.js` is not a module

**Review by wbinnssmith — APPROVED** [review]

**Pull request merged**